drivers/display/display_nrf_led_matrix.c

```diff
@@ -0,0 +1,448 @@
+/*
+ * Copyright (c) 2021, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/display.h>
+#include <devicetree.h>
+#include <dt-bindings/gpio/gpio.h>
+#include <hal/nrf_timer.h>
+#ifdef PWM_PRESENT
+#include <hal/nrf_pwm.h>
+#endif
+#include <nrfx_gpiote.h>
+#ifdef PPI_PRESENT
+#include <nrfx_ppi.h>
+#endif
+
+#define MATRIX_NODE  DT_INST(0, nrf_led_matrix)
+#define TIMER_NODE   DT_PHANDLE(MATRIX_NODE, timer)
+#define USE_PWM      DT_NODE_HAS_PROP(MATRIX_NODE, pwm)
+#define ROW_COUNT    DT_PROP_LEN(MATRIX_NODE, row_gpios)
+#define COL_COUNT    DT_PROP_LEN(MATRIX_NODE, col_gpios)
+
+#define X_PIXELS     DT_PROP(MATRIX_NODE, width)
+#define Y_PIXELS     DT_PROP(MATRIX_NODE, height)
+#define PIXEL_COUNT  DT_PROP_LEN(MATRIX_NODE, pixel_mapping)
+BUILD_ASSERT(PIXEL_COUNT == (X_PIXELS * Y_PIXELS),
+	     "Invalid length of pixel-mapping");
+
+#define PIXEL_MAPPING(idx)  DT_PROP_BY_IDX(MATRIX_NODE, pixel_mapping, idx)
+#define CHECK_PIXEL(node_id, pha, idx) \
+	BUILD_ASSERT((PIXEL_MAPPING(idx) >> 4) < ROW_COUNT, \
+		     "Invalid row index in pixel-mapping["#idx"]"); \
+	BUILD_ASSERT((PIXEL_MAPPING(idx) & 0xF) < COL_COUNT, \
+		     "Invalid column index in pixel-mapping["#idx"]");
+DT_FOREACH_PROP_ELEM(MATRIX_NODE, pixel_mapping, CHECK_PIXEL)
+
+#define BASE_FREQUENCY     16000000
+#define TIMER_CLK_CONFIG   NRF_TIMER_FREQ_16MHz
+#define PWM_CLK_CONFIG     NRF_PWM_CLK_16MHz
+#define REFRESH_FREQUENCY  50
+#define BRIGHTNESS_MAX     255
+
+#define QUANTUM  (BASE_FREQUENCY / (REFRESH_FREQUENCY * PIXEL_COUNT * \
+				    BRIGHTNESS_MAX))
+#define PIXEL_PERIOD  (BRIGHTNESS_MAX * QUANTUM)
+BUILD_ASSERT(PIXEL_PERIOD <= BIT_MASK(16));
+
+#define ACTIVE_LOW_MASK  0x80
+#define PSEL_MASK        0x7F
+
+struct display_drv_config {
+	NRF_TIMER_Type *timer;
+#if USE_PWM
+	NRF_PWM_Type *pwm;
+#endif
+	uint8_t rows[ROW_COUNT];
+	uint8_t cols[COL_COUNT];
+	uint8_t pixel_mapping[PIXEL_COUNT];
+};
+
+struct display_drv_data {
+#if USE_PWM
+	uint16_t seq;
+#else
+	uint8_t gpiote_ch;
+#endif
+	uint8_t pixel_idx;
+	uint8_t framebuf[PIXEL_COUNT];
+	uint8_t brightness;
+	bool    blanking;
+};
+
+static void set_pin(uint8_t pin_info, bool active)
+{
+	uint32_t value = active ? 1 : 0;
+
+	if (pin_info & ACTIVE_LOW_MASK) {
+		value = !value;
+	}
+	nrf_gpio_pin_write(pin_info & PSEL_MASK, value);
+}
+
+static int api_blanking_on(const struct device *dev)
+{
+	struct display_drv_data *dev_data = dev->data;
+	const struct display_drv_config *dev_config = dev->config;
```

Comment:
`data`, `config` are usual var names

Reply:
I find them too generic and I had cases in the past where some other `config` or `data` appeared in a given function. So to avoid confusion I'd prefer to keep these prefixes (these are only 4 extra characters). Many nRF drivers actually use such naming.